File: frontend/src/types/license.ts
export type LicenseTier = 'community' | 'pro' | 'enterprise';

export interface LicenseInfo {
  tier: LicenseTier;
  valid: boolean;
  max_hosts: number;
  current_hosts: number;
  features: string[];
  expiry: string | null;
  grace_period: boolean;
  grace_period_end: string | null;
  last_validated: string;
  error: string;
  license_key_masked: string;
}

export const TIER_LABELS: Record<LicenseTier, string> = {
  community: 'Community',
  pro: 'Pro',
  enterprise: 'Enterprise',
};

export const TIER_COLORS: Record<LicenseTier, string> = {
  community: 'text-slate-500 bg-slate-500/10 border-slate-500/20',
  pro: 'text-cyan-500 bg-cyan-500/10 border-cyan-500/20',
  enterprise: 'text-purple-500 bg-purple-500/10 border-purple-500/20',
};

export const FEATURE_LABELS: Record<string, string> = {
  stacks: 'Stack Management',
  alerts: 'Alert System',
  multi_host: 'Multi-Host',
  update_check: 'Image Update Checking',
  rbac: 'Role-Based Access Control',
  sso: 'Single Sign-On',
  audit_logs: 'Audit Logs',
};

export const ALL_FEATURES = ['stacks', 'alerts', 'multi_host', 'update_check', 'rbac', 'sso', 'audit_logs'];

File: backend/internal/models/license.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type LicenseTier string

const (
	TierCommunity  LicenseTier = "community"
	TierPro        LicenseTier = "pro"
	TierEnterprise LicenseTier = "enterprise"
)

// LicenseCache stores the last known license state in the database.
// Uses a singleton pattern (single row, ID=1) via FirstOrCreate.
type LicenseCache struct {
	gorm.Model
	LicenseKey     string      `gorm:"size:512"`
	Tier           LicenseTier `gorm:"size:32;default:community"`
	Valid          bool        `gorm:"default:false"`
	Expiry         *time.Time
	MaxHosts       int    `gorm:"default:1"`
	Features       string `gorm:"type:text"` // JSON array of feature codes
	MachineID            string `gorm:"size:256"` // Server fingerprint (hostname + MAC)
	LicenciaActivationID string `gorm:"size:256"` // Reserved for future Licencia activation tracking
	LastValidated  time.Time
	LastError      string `gorm:"type:text"`
	GracePeriodEnd *time.Time
}

// DefaultFeatures returns the hardcoded feature set for a given tier.
func DefaultFeatures(tier LicenseTier) []string {
	switch tier {
	case TierEnterprise:
		return []string{"stacks", "alerts", "multi_host", "update_check", "rbac", "sso", "audit_logs"}
	case TierPro:
		return []string{"stacks", "alerts", "multi_host", "update_check"}
	default:
		return []string{}
	}
}

// DefaultMaxHosts returns the default host limit for a given tier.
func DefaultMaxHosts(tier LicenseTier) int {
	switch tier {
	case TierEnterprise:
		return -1 // unlimited
	case TierPro:
		return 10
	default:
		return 1
	}
}
